rasterio/_io.pyx

```diff
@@ -1658,9 +1658,11 @@ cdef class InMemoryRaster:
                        count, <GDALDataType>dtypes.dtype_rev[dtype], NULL))
 
         if transform is not None:
+            self.transform = transform
+            transform = transform.to_gdal()
```

Comment:
@sgillies suggestion:
`gdal_transform = transform.to_gdal()` to avoid any ambiguity and keep `transform` always as an `Affine`